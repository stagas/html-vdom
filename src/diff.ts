import { VList } from './v'

// import { inspect } from 'util'
// const log = (...args: any[]) => console.log(...args.map(x => inspect(x, false, Infinity, true)))

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ParentLike = ParentNode | VList<any>

const $ = (P: ParentLike) => (Array.isArray(P) ? P.at(-1) : P.lastChild)

// const nextSibling = <T>(parent: ParentLike<T>, child: T | ChildNode): T | ChildNode | null =>
//   parent instanceof VList ? parent.nextSiblingOf(child as T) : (child as ChildNode).nextSibling

const _ = <T>(parent: ParentLike | T[], index: number): T | null =>
  Array.isArray(parent)
    ? (parent[index] as T) ?? null
    : ((parent as ParentNode).childNodes[index] as unknown as T) ?? null

const len = (P: ParentLike) => (Array.isArray(P) ? P.length : P.childNodes.length)

export const diff = <T>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  eq: (a: any, b: any) => boolean,
  P: ParentLike,
  b: T[],
  a: T[] = []
) => {
  // log(a, b)

  // A holds the anchor element where insertBefore(x, A) latches on
  let A: T | null

  // B holds the current element in the new list b
  let B: T | null

  // C holds a temp value of the list a that will be compared and maybe inserted
  let C: T | null

  // log(b, a)
  let i = 0

  const n = b.length
  let p = a.length

  //              p
  // a  1 2 3 4 5
  //        i
  // b  1 2 0 4 5
  //              n
  // R  1 2
  //
  // advance start pointer while equal head
  // TODO:                     set A B here
  while (i < n && i < p && eq((A = _(a, i)), (B = _(b, i)))) i++

  //          p
  // a  1 2 3 4 5
  //        i
  // b  1 2 0 4 5
  //          n
  // R  1 2
  //
  // shorten end pointer while equal tail
  // while (i < n && i < p && eq(_(a, p - 1), _(b, n - 1))) --p, --n

  // insert:
  //        A
  //        p
  // a  1 2 4 5
  //              i
  // b  1 2 0 0 0 4 5
  //              n
  // R  1 2 0 0 0
  //

  // append:
  //          A
  //          p
  // a  1 2 3
  //          i
  // b  1 2 3 4
  //            n
  //

  // prepend:
  //
  //    A
  //    p
  // a  1 2 3
  //    i
  // b  4 1 2 3
  //      n
  //

  // insert/append/prepend shortcut
  if (i == p) {
    A = _(a, p)
    for (; i < n; i++) P.insertBefore(_(b, i)!, A)
    // we are done -- rest of elements (if any) are the same
  } else {
    // replace:
    //
    //      A
    //        p
    // a  1 2 3
    //      i
    // b  1 4 3
    //        n
    //      B

    // replace + remaining tail:
    //
    //        A
    //        p
    // a  1 2 3
    //        i
    // b  1 4 5 6 3
    //            n
    // R  1 4 3
    //      B

    //      x
    //      A
    //          p
    // a  1 2 3 4
    //        i
    //        C
    //            y
    // b  1 3 2 4
    //          n
    //            B
    // R  1 3 2 4

    //      x
    //      A
    //            p
    // a  1 2 3 4
    //        C
    //        i
    //      y
    // b  1 3 4 2
    //      B
    //            n
    // R  1 3 4 2

    // moves:
    //
    //        x
    //                    p
    //        A
    // a  1 2 3 4 5 6 7 8
    //    i         y
    //    C
    // b  3 6 5 1 2 7 8 4
    //              B
    //                    n
    //
    // R  1 2 3 4 5 6 7 8
    // R  3 6 5 1 2 4 7 8
    //

    // let x = i
    // let y = i
    // for (; y < n; y++) {
    //   for (i = y; i < p; i++) {
    //     C = _(P, i)
    //     if (eq(C, B)) {
    //       P.insertBefore(C, A)
    //       break
    //     }
    //     //
    //   }

    //   while (y < n && x < p && eq((A = _(P, y)), (B = _(b, y)))) x++, y++

    //   //      A = _(P, x)
    // }

    let x = 0
    out: while (((A = _(P, i)!), (B = _(b, i)!), i++ < n)) {
      if (!eq(A, B)) {
        for (x = i; x < p; x++) {
          C = _(P, x)!
          if (eq(B, C)) {
            P.insertBefore(C, A)
            continue out
          }
        }
        for (x = i; x < n; x++) {
          C = _(b, x)
          if (eq(A, C)) {
            P.insertBefore(B, A)
            continue out
          }
        }
        if (A) P.replaceChild(B, A)
        else P.appendChild(B) //, null)
      }
    }
    p = len(P)
    i = n
    while (i++ < p) P.removeChild($(P))

    // remove tail:
    //
    //          p
    // a  1 2 3
    //      i
    // b  1 2
    //        n

    // remove head:
    //
    //      p
    // a  1 2 3
    //    i
    // b  2 3
    //    n
    // for (; i < p--; ) P.removeChild(_(P, i))
  }

  // if (P instanceof Node) {
  //   log('RESULT', P.innerHTML)
  // } else log('RESULT', [...P])
}
