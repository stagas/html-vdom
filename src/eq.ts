import { VAtom } from '.'

export const atomEq = (a: VAtom, b: VAtom) => {
  // prettier-ignore
  const eq = (
    (a != null && b != null) &&
    (
      (a === b) ||
      (
        (a[1] && b[1] && 'key' in a[1] && 'key' in b[1])
          ? a[1]!.key === b[1]!.key
          : a[0] === b[0]
      )
    )
  )
  // console.log('EQUAL', eq, a, b)
  return eq
}

export const domEq = (a: Node, b: Node) => {
  return a === b
}
