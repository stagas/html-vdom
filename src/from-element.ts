import * as jsxi from 'html-jsx'
import { jsx } from './jsx-runtime'

export type Constructor<T> = new(...args: any[]) => T

const Registry: WeakMap<typeof HTMLElement, any> = new WeakMap()

export interface Options<T> extends ElementDefinitionOptions {
  interface?: Constructor<T>
}

export type Component<T, I> = (props: Partial<Omit<T & I, keyof HTMLElement>> & jsxi.HTMLAttributes<T>) => any

export const fromElement = <T extends HTMLElement, I = HTMLElement>(
  Element: Constructor<T>,
  options: Options<I> = {},
): Component<T, I> => {
  let fn: Component<T, I> = Registry.get(Element)
  if (fn) return fn

  const tag = 'x-' + Element.name.toLowerCase() + ((Math.random() * 1e7) | 0).toString(36)
  const ctor = class extends Element {}
  customElements.define(tag, ctor, options)
  if (options.extends)
    fn = props => jsx(options.extends, Object.assign(props, { is: tag }), void 0)
  else
    fn = props => jsx(tag, props, void 0)
  Registry.set(Element, fn)
  ;(fn as any).toString = () => tag

  return fn
}
