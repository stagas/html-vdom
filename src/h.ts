/**
 * JSX types
 * @private
 */

import type * as jsx from 'html-jsx'
declare module 'html-jsx' {
  // eslint-disable-next-line
  interface DOMAttributes<T> extends JSX.IntrinsicAttributes {}
}
declare global {
  namespace JSX {
    /**
     * The type returned by our `createElement` factory.
     * @private
     */
    type Element = VAtom

    interface IntrinsicElements extends jsx.IntrinsicElements {
      /**
       * This allows for any tag to be used.
       * @private
       */
      [k: string]: unknown
    }

    interface IntrinsicAttributes {
      /**
       * Pass an object that will be assigned the DOM element reference in its `current` property.
       * @private
       */
      ref?: VRef<unknown> | null

      /**
       * List index key - each item's `key` must be unique.
       * @private
       */
      key?: string | number
    }

    interface HTMLAttributes<T> extends jsx.HTMLAttributes<T> {}
    interface SVGAttributes<T> extends jsx.SVGAttributes<T> {}
    interface DOMAttributes<T> extends jsx.DOMAttributes<T> {}
  }
}

import { VAtom as VAtomCtor, VList } from './v'

export interface VRef<T> {
  current?: T | null
}

export const Fragment = Symbol()
export type JSXFragment = typeof Fragment
export type JSXKind = string | JSXFragment | JSXFunction
export type JSXProps = (Record<string, unknown> & { key?: string | number }) | null
export type JSXChild = JSXReturn[] | JSXReturn | string | number | boolean | null | undefined
export type JSXReturn = [JSXKind, JSXProps, VList<JSXChild>]
export interface JSXFunction {
  (props?: JSXProps & { children?: JSXChild[] }): JSXReturn
}
export interface JSXPragma {
  (kind: JSXKind, props: JSXProps, ...children: JSXChild[]): JSXReturn //VAtom | VList<VAtom>
}

export const h: JSXPragma = (kind, props, ...children) =>
  new VAtomCtor(kind, props, new VList(...(children.map(toAtom) as VAtom[]))) as VAtom

export type VAtom = [JSXKind, JSXProps, VList<VAtom>]

const toAtom = (child: JSXChild) => {
  if (child == null)
    //return // new VAtomCtor()
    return new VAtomCtor(<never>Comment) as VAtom

  switch (typeof child) {
    case 'boolean':
      return new VAtomCtor(<never>Comment) as VAtom
    // return false //new VAtomCtor() //return new VAtomCtor(<never>Comment)
    case 'string':
    case 'number':
      return new VAtomCtor(<never>Text, null, <never>child) as VAtom
    case 'object':
      if (Array.isArray(child) && !(child instanceof VAtomCtor))
        return h(Fragment, null, ...(child as JSXReturn[]))
    //Array === child?.constructor) return h(Fragment, null, ...(child as JSXReturn[]))
  }

  return child
}
