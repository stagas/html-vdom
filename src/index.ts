export * from './h'
export * from './render'
