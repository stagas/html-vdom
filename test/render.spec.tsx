import { Fragment, VHook, current, h, render } from '../src'

let c: HTMLDivElement

const html = () => c.innerHTML
const $ = (sel: string) => c.querySelector(sel)!
const $$ = (sel: string) => c.querySelectorAll(sel)!

beforeEach(() => {
  c = document.createElement('div')
})

describe('render', () => {
  it('single div', () => {
    render(<div></div>, c)
    expect(html()).toMatchSnapshot()
  })

  it('morphs div to p', () => {
    render(<div></div>, c)
    expect(html()).toMatchSnapshot()

    render(<p></p>, c)
    expect(html()).toMatchSnapshot()
  })

  it('with child text', () => {
    render(<div>hello world</div>, c)
    expect(html()).toMatchSnapshot()
  })

  it('props create', () => {
    render(<div id="foo">hello world</div>, c)
    expect(html()).toMatchSnapshot()
  })

  it('props update', () => {
    render(<div id="foo">hello world</div>, c)
    expect(html()).toMatchSnapshot()
    render(<div id="bar">hello world</div>, c)
    expect(html()).toMatchSnapshot()
  })

  it('morph same lane', () => {
    render(
      <>
        hello
        <> there </>
        world
      </>,
      c
    )
    expect(html()).toMatchSnapshot()

    render(
      <>
        hello
        <div> there </div>
        world
      </>,
      c
    )
    expect(html()).toMatchSnapshot()
  })

  it('morphs child text keeping references', () => {
    render(<div>hello</div>, c)
    expect(html()).toMatchSnapshot()

    const el = $('div')
    const text = el.firstChild

    render(<div>world</div>, c)
    expect(html()).toMatchSnapshot()

    expect($('div')).toBe(el)
    expect($('div').firstChild).toBe(text)
  })

  it('ul multiple children', () => {
    render(
      <ul>
        <li key="a">a</li>
        <li key="b">b</li>
        <li key="c">c</li>
      </ul>,
      c
    )
    expect(html()).toMatchSnapshot()

    const lis = $$('li')

    render(
      <ul>
        <li key="c">c</li>
        <li key="a">a</li>
        <li key="b">b</li>
      </ul>,
      c
    )
    expect(html()).toMatchSnapshot()

    const res = $$('li')

    expect(res[1]).toBe(lis[0])
    expect(res[2]).toBe(lis[1])
    expect(res[0]).toBe(lis[2])
  })

  it('changing children multiple times', () => {
    render(
      <ul>
        <li key="a">a</li>
        <li key="b">b</li>
        <li key="c">c</li>
      </ul>,
      c
    )
    expect(html()).toMatchSnapshot()

    {
      const lis = $$('li')

      render(
        <ul>
          <li key="c">c</li>
          <li key="a">a</li>
          <li key="b">b</li>
        </ul>,
        c
      )
      expect(html()).toMatchSnapshot()

      const res = $$('li')

      expect(res[1]).toBe(lis[0])
      expect(res[2]).toBe(lis[1])
      expect(res[0]).toBe(lis[2])
    }

    {
      const lis = $$('li')

      render(
        <ul>
          <li key="a">a</li>
          <li key="d">d</li>
          <li key="c">c</li>
          <li key="b">b</li>
        </ul>,
        c
      )
      expect(html()).toMatchSnapshot()

      const res = $$('li')

      expect(res[0]).toBe(lis[1])
      expect(res[2]).toBe(lis[0])
      expect(res[3]).toBe(lis[2])
    }

    {
      const lis = $$('li')

      render(
        <ul>
          <li key="d">d</li>
          <li key="b">b</li>
        </ul>,
        c
      )
      expect(html()).toMatchSnapshot()

      const res = $$('li')

      expect(res[0]).toBe(lis[1])
      expect(res[1]).toBe(lis[3])
    }

    {
      render(<ul></ul>, c)
      expect(html()).toMatchSnapshot()
    }
  })

  it('complex tree', () => {
    render(
      <div>
        <ul>
          <li key="a">
            <button>
              some text <span>a</span>
            </button>
          </li>
          <li key="b">
            <button>
              here too <span>b</span>
            </button>
          </li>
          <li key="c">
            <button>
              <span>c</span> and here
            </button>
          </li>
        </ul>
      </div>,
      c
    )
    expect(html()).toMatchSnapshot()

    const bs = $$('button')

    render(
      <div>
        <ul>
          <li key="a">
            <button>
              some text <span>a</span>
            </button>
          </li>
          <li key="c">
            <button>
              <span>c</span> and here
            </button>
          </li>
          <li key="d">
            <button>
              hey <span>d</span> yo
            </button>
          </li>
          <li key="b">
            <button>
              here too <span>b</span>
            </button>
          </li>
        </ul>
      </div>,
      c
    )
    expect(html()).toMatchSnapshot()

    const res = $$('button')

    expect(res[1]).toBe(bs[2])
    expect(res[3]).toBe(bs[1])
    expect(res[0]).toBe(bs[0])
  })

  describe('fragment', () => {
    it('simple text', () => {
      render(<>hello world</>, c)
      expect(html()).toMatchSnapshot()
    })

    it('changing text', () => {
      render(<>hello world</>, c)
      expect(html()).toMatchSnapshot()

      render(<>another world</>, c)
      expect(html()).toMatchSnapshot()
    })

    it('multiple children', () => {
      render(
        <>
          hello
          <div>world</div>
          <span>hey</span>
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
    })

    it('changing and reusing children', () => {
      render(
        <>
          hello
          <div>world</div>
          <span>hey</span>
        </>,
        c
      )
      expect(html()).toMatchSnapshot()

      const els = $$('*')

      render(
        <>
          <div>world</div>
          <div>other</div>
          another
          <span>yo</span>
        </>,
        c
      )
      expect(html()).toMatchSnapshot()

      const res = $$('*')

      expect(res[0]).toBe(els[0])
      expect(res[2]).toBe(els[1])
    })

    it('fragments within fragments', () => {
      render(
        <>
          <>one</>
          <>
            hello
            <div>world</div>
          </>
          <span>
            <>hey</>
          </span>
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
    })

    it('changing fragments within fragments', () => {
      render(
        <>
          <>one</>
          <>
            hello
            <div>world</div>
          </>
          <span>
            <>hey</>
          </span>
        </>,
        c
      )
      expect(html()).toMatchSnapshot()

      render(
        <>
          <>
            hello
            <div>world</div>
          </>
          <span>
            <>hey</>
          </span>
          <>one</>
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
    })
  })

  describe('arrays', () => {
    it('renders', () => {
      render(
        <ul>
          {'abc'.split('').map(x => (
            <li key={x}>{x}</li>
          ))}
        </ul>,
        c
      )
      expect(html()).toMatchSnapshot()
    })

    it('reorders', () => {
      render(
        <ul>
          {'abc'.split('').map(x => (
            <li key={x}>{x}</li>
          ))}
        </ul>,
        c
      )
      expect(html()).toMatchSnapshot()

      const lis = $$('li')

      render(
        <ul>
          {'cab'.split('').map(x => (
            <li key={x}>{x}</li>
          ))}
        </ul>,
        c
      )
      expect(html()).toMatchSnapshot()

      const res = $$('li')

      expect(res[1]).toBe(lis[0])
      expect(res[2]).toBe(lis[1])
      expect(res[0]).toBe(lis[2])
    })

    it('fragments', () => {
      render(
        <ul>
          {'abc'.split('').map(x => (
            <>
              <li>{x}</li>
            </>
          ))}
        </ul>,
        c
      )
      expect(html()).toMatchSnapshot()
    })
  })

  describe('functions', () => {
    it('empty fragment', () => {
      const Foo = () => <></>
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
    })

    it('simple text', () => {
      const Foo = () => <>hello world</>
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
    })

    it('with children', () => {
      const Foo = ({ children }: { children?: never }) => <>{children}</>
      render(<Foo>hello world</Foo>, c)
      expect(html()).toMatchSnapshot()
    })

    it('pass props', () => {
      const Foo = ({ answer }: { answer: number }) => (
        <>
          answer is <b>{answer}</b>
        </>
      )
      render(<Foo answer={42} />, c)
      expect(html()).toMatchSnapshot()
    })
  })

  describe('hooks', () => {
    it('trigger', () => {
      let i = 0
      let hook!: VHook
      const Foo = () => {
        hook = current.hook!
        return <>{i++}</>
      }
      render(
        <>
          <Foo />
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
    })

    it('mulitple triggers', () => {
      let i = 0
      let hook!: VHook
      const Foo = () => {
        hook = current.hook!
        return <>{i++}</>
      }
      render(
        <>
          <Foo />
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
    })

    it('they dont leak', () => {
      const set: Set<VHook> = new Set()
      let i = 0
      let hook!: VHook
      const Foo = () => {
        hook = current.hook!
        set.add(hook)
        return <>{i++}</>
      }
      render(
        <>
          <Foo />
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
      set.forEach(x => x.trigger())
      expect(html()).toMatchSnapshot()
      set.forEach(x => x.trigger())
      expect(html()).toMatchSnapshot()
      set.forEach(x => x.trigger())
      expect(html()).toMatchSnapshot()
      expect(set.size).toEqual(1)
    })

    it('deep they dont leak', () => {
      const set: Set<VHook> = new Set()
      let i = 0
      let hook!: VHook
      let hook_foo!: VHook
      const Bar = () => {
        hook = current.hook!
        set.add(hook)
        return <>{i++}</>
      }
      const Foo = () => {
        hook = current.hook!
        hook_foo = hook
        set.add(hook)
        return <Bar />
      }
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
      expect(set.size).toEqual(2)
      hook_foo.trigger()
      // set.forEach(x => x.trigger())
      expect(html()).toMatchSnapshot()
      expect(set.size).toEqual(2)
      hook_foo.trigger()
      // set.forEach(x => x.trigger())
      expect(html()).toMatchSnapshot()
      expect(set.size).toEqual(2)
      // set.forEach(x => x.trigger())
      // expect(html()).toMatchSnapshot()
      // expect(set.size).toEqual(2)
    })

    it('updates only self', () => {
      let i = 0
      let hook!: VHook
      const Foo = () => {
        hook = current.hook!
        return <>{i++}</>
      }
      render(
        <>
          hello
          <Foo />
          world
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
    })

    it('changes top element', () => {
      let i = 0
      let hook!: VHook
      const Foo = () => {
        hook = current.hook!
        return i % 2 === 0 ? <>{i++}</> : <div>{i++}</div>
      }
      render(
        <>
          hello
          <Foo />
          world
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
    })

    it('changes deeply element', () => {
      let i = 0
      let hook_bar!: VHook
      const Bar = () => {
        hook_bar = current.hook!
        return <>{i++}</>
      }
      const Foo = () => {
        return (
          <main>
            <div>
              <Bar />
            </div>
          </main>
        )
      }
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
      hook_bar.trigger()
      expect(html()).toMatchSnapshot()
    })

    it('changes deeply with children', () => {
      let i = 41
      let hook_bar!: VHook
      const Bar = ({ children }: { children?: never }) => {
        hook_bar = current.hook!
        return (
          <>
            {children} {i++}
          </>
        )
      }
      const Foo = () => {
        return (
          <main>
            <div>
              <Bar>
                <span>the answer is</span>
              </Bar>
            </div>
          </main>
        )
      }
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
      // render(<Foo />, c)
      hook_bar.trigger()
      expect(html()).toMatchSnapshot()
    })
  })
})

describe('quirky cases', () => {
  it('pass component in props', () => {
    let i = 0
    const Bar = ({ inner }) => (
      <p>
        {i++} {inner}
      </p>
    )
    const Foo = ({ Component, children }: { Component: any; children?: any }) => (
      <Component inner={children} />
    )
    render(<Foo Component={Bar}>hello</Foo>, c)
    expect(html()).toMatchSnapshot()
  })

  it('renders input elements correctly', () => {
    render(
      <>
        <input type="range" min="5" max="10" value="8" />
        <input type="number" min="5" max="10" value="8" />
      </>,
      c
    )
    expect(html()).toMatchSnapshot()
  })
})
