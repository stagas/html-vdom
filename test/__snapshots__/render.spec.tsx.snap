// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all quirky cases 2 keyed removed and added again 1`] = `"<!---->"`;

exports[`all quirky cases 2 keyed removed and added again 2`] = `"<p>a</p><p>b</p><p>c</p><p>d</p><p>e</p>"`;

exports[`all quirky cases 2 keyed removed and added again 3`] = `"<p>a</p><p>f</p><p>b</p><p>c</p>"`;

exports[`all quirky cases keyed removed and added again 1`] = `"<!---->"`;

exports[`all quirky cases keyed removed and added again 2`] = `"<p>a</p><p>b</p><p>c</p><p>d</p>"`;

exports[`all quirky cases keyed removed and added again 3`] = `"<p>a</p><p>c</p>"`;

exports[`all quirky cases keyed removed and added again 4`] = `"<p>a</p><p>b</p><p>c</p><p>d</p><p>e</p>"`;

exports[`all quirky cases keyed removed and added again 5`] = `"<p>d</p><p>e</p>"`;

exports[`all quirky cases keyed removed and added again 6`] = `"<p>e</p><p>b</p><p>c</p>"`;

exports[`all quirky cases keyed removed and added again 7`] = `"<p>d</p><p>c</p><p>e</p>"`;

exports[`all quirky cases less then many then less 1`] = `"<p>0</p><p>2</p><p>4</p><p>1</p><p>3</p>"`;

exports[`all quirky cases less then many then less 2`] = `"<p>0</p><p>2</p><p>4</p><p>6</p><p>8</p><p>1</p><p>3</p><p>5</p><p>7</p><p>9</p>"`;

exports[`all quirky cases less then many then less 3`] = `"<p>0</p><p>2</p><p>1</p><p>3</p>"`;

exports[`all quirky cases less then many then less 4`] = `"<p>0</p><p>2</p><p>4</p><p>6</p><p>8</p><p>1</p><p>3</p><p>5</p><p>7</p><p>9</p>"`;

exports[`all quirky cases pass component in props 1`] = `"<p>0 hello</p>"`;

exports[`all quirky cases render children at same position properly 1`] = `"<div>hello 1</div>"`;

exports[`all quirky cases render children at same position properly 2`] = `"<div>hello 2</div>"`;

exports[`all quirky cases render children at same position properly 3`] = `"<div>hello 3</div>"`;

exports[`all quirky cases renders input elements correctly 1`] = `"<input type=\\"range\\" min=\\"5\\" max=\\"10\\"><input type=\\"number\\" min=\\"5\\" max=\\"10\\">"`;

exports[`all render arrays fragments 1`] = `"<ul><li>a</li><li>b</li><li>c</li></ul>"`;

exports[`all render arrays renders 1`] = `"<ul><li>a</li><li>b</li><li>c</li></ul>"`;

exports[`all render arrays reorders 1`] = `"<ul><li>a</li><li>b</li><li>c</li></ul>"`;

exports[`all render arrays reorders 2`] = `"<ul><li>c</li><li>a</li><li>b</li></ul>"`;

exports[`all render changing children multiple times 1`] = `"<ul><li>a</li><li>b</li><li>c</li></ul>"`;

exports[`all render changing children multiple times 2`] = `"<ul><li>c</li><li>a</li><li>b</li></ul>"`;

exports[`all render changing children multiple times 3`] = `"<ul><li>a</li><li>d</li><li>c</li><li>b</li></ul>"`;

exports[`all render changing children multiple times 4`] = `"<ul><li>d</li><li>b</li></ul>"`;

exports[`all render changing children multiple times 5`] = `"<ul></ul>"`;

exports[`all render complex tree 1`] = `"<div><ul><li><button>some text <span>a</span></button></li><li><button>here too <span>b</span></button></li><li><button><span>c</span> and here</button></li></ul></div>"`;

exports[`all render complex tree 2`] = `"<div><ul><li><button>some text <span>a</span></button></li><li><button><span>c</span> and here</button></li><li><button>hey <span>d</span> yo</button></li><li><button>here too <span>b</span></button></li></ul></div>"`;

exports[`all render component conditional swap ref 1`] = `"<div>hello world</div>"`;

exports[`all render component conditional swap ref 2`] = `"<p>other element</p>"`;

exports[`all render component conditional swap ref 3`] = `"<div>hello world</div>"`;

exports[`all render conditional swap raf 1`] = `"<div>hello world</div>"`;

exports[`all render conditional swap raf 2`] = `"<p>other element</p>"`;

exports[`all render conditional swap raf 3`] = `"<div>hello world</div>"`;

exports[`all render fragment changing fragments within fragments 1`] = `"onehello<div>world</div><span>hey</span>"`;

exports[`all render fragment changing fragments within fragments 2`] = `"hello<div>world</div><span>hey</span>one"`;

exports[`all render fragment changing text 1`] = `"hello world"`;

exports[`all render fragment changing text 2`] = `"another world"`;

exports[`all render fragment fragments within fragments 1`] = `"onehello<div>world</div><span>hey</span>"`;

exports[`all render fragment multiple children 1`] = `"hello<div>world</div><span>hey</span>"`;

exports[`all render fragment simple text 1`] = `"hello world"`;

exports[`all render functions empty fragment 1`] = `"<!---->"`;

exports[`all render functions pass props 1`] = `"answer is <b>42</b>"`;

exports[`all render functions simple text 1`] = `"hello world"`;

exports[`all render functions with children 1`] = `"hello world"`;

exports[`all render hooks changes deeply element 1`] = `"<main><div>0</div></main>"`;

exports[`all render hooks changes deeply element 2`] = `"<main><div>1</div></main>"`;

exports[`all render hooks changes deeply with children 1`] = `"<main><div><span>the answer is</span> 41</div></main>"`;

exports[`all render hooks changes deeply with children 2`] = `"<main><div><span>the answer is</span> 42</div></main>"`;

exports[`all render hooks changes top element 1`] = `"hello0world"`;

exports[`all render hooks changes top element 2`] = `"hello<div>1</div>world"`;

exports[`all render hooks changes top element 3`] = `"hello2world"`;

exports[`all render hooks changes top element 4`] = `"hello<div>3</div>world"`;

exports[`all render hooks deep they dont leak 1`] = `"0"`;

exports[`all render hooks deep they dont leak 2`] = `"1"`;

exports[`all render hooks deep they dont leak 3`] = `"2"`;

exports[`all render hooks mulitple triggers 1`] = `"0"`;

exports[`all render hooks mulitple triggers 2`] = `"1"`;

exports[`all render hooks mulitple triggers 3`] = `"2"`;

exports[`all render hooks mulitple triggers 4`] = `"3"`;

exports[`all render hooks they dont leak 1`] = `"0"`;

exports[`all render hooks they dont leak 2`] = `"1"`;

exports[`all render hooks they dont leak 3`] = `"2"`;

exports[`all render hooks they dont leak 4`] = `"3"`;

exports[`all render hooks trigger 1`] = `"0"`;

exports[`all render hooks trigger 2`] = `"1"`;

exports[`all render hooks updates only self 1`] = `"hello0world"`;

exports[`all render hooks updates only self 2`] = `"hello1world"`;

exports[`all render hooks updates only self 3`] = `"hello2world"`;

exports[`all render hooks updates only self 4`] = `"hello3world"`;

exports[`all render morph same lane 1`] = `"hellothereworld"`;

exports[`all render morph same lane 2`] = `"hello<div>there</div>world"`;

exports[`all render morphs child text keeping references 1`] = `"<div>hello</div>"`;

exports[`all render morphs child text keeping references 2`] = `"<div>world</div>"`;

exports[`all render morphs div to p 1`] = `"<div></div>"`;

exports[`all render morphs div to p 2`] = `"<p></p>"`;

exports[`all render props create 1`] = `"<div id=\\"foo\\">hello world</div>"`;

exports[`all render props update 1`] = `"<div id=\\"foo\\">hello world</div>"`;

exports[`all render props update 2`] = `"<div id=\\"bar\\">hello world</div>"`;

exports[`all render ref & pass ref 1`] = `"<div>hello world</div>"`;

exports[`all render ref & pass ref 2`] = `"<div>same element</div>"`;

exports[`all render single div 1`] = `"<div></div>"`;

exports[`all render ul multiple children 1`] = `"<ul><li>a</li><li>b</li><li>c</li></ul>"`;

exports[`all render ul multiple children 2`] = `"<ul><li>c</li><li>a</li><li>b</li></ul>"`;

exports[`all render with child text 1`] = `"<div>hello world</div>"`;

exports[`all return same reference add + shuffle 1`] = `"123"`;

exports[`all return same reference add + shuffle 2`] = `"2413"`;

exports[`all return same reference remove + shuffle 1`] = `"1234"`;

exports[`all return same reference remove + shuffle 2`] = `"31"`;

exports[`all return same reference shuffle 1`] = `"123"`;

exports[`all return same reference shuffle 2`] = `"213"`;

exports[`all svg after hook update cached contents 1`] = `"<svg viewBox=\\"0 0 10 10\\"><!----></svg>"`;

exports[`all svg after hook update cached contents 2`] = `"<svg viewBox=\\"0 0 10 10\\"><path d=\\"M 0,20 h100\\" pathLength=\\"100\\"></path><path d=\\"M 0,20 h100\\" pathLength=\\"100\\"></path></svg>"`;

exports[`all svg svg 1`] = `"<svg></svg>"`;

exports[`all svg svg then foreignObject 1`] = `"<svg viewBox=\\"0 0 10 10\\"><path pathLength=\\"90\\"></path><foreignObject><input></foreignObject></svg>"`;

exports[`all svg svg then foreignObject right attribute casing 1`] = `"<svg viewBox=\\"0 0 10 10\\"><path pathLength=\\"90\\"></path><foreignObject><input autofocus=\\"\\"></foreignObject></svg>"`;

exports[`all svg uses correct namespace 1`] = `"<svg viewBox=\\"0 0 10 10\\"></svg>"`;

exports[`all svg uses correct namespace after hook update 1`] = `"<div><svg viewBox=\\"0 0 10 10\\"><path d=\\"M 0,20 h100\\" pathLength=\\"50\\"></path></svg></div>"`;

exports[`all svg uses correct namespace after hook update 2`] = `"<div><svg viewBox=\\"0 0 10 10\\"><path d=\\"M 0,20 h100\\" pathLength=\\"100\\"></path><path d=\\"M 0,20 h100\\" pathLength=\\"100\\"></path></svg></div>"`;
